### PR TITLE
Denormalize from_url_without_rest onto the redirects model

### PR DESCRIPTION
**Pull request by ericholscher**
This should make querying faster and the logic to support it easier

Refs https://github.com/readthedocs/readthedocs.org/pull/6779

### CODE REVIEW

**Comment by ericholscher** [comment]
@humitos I think this is mostly ready, and we should be able to ship it if we want 🤞

**Comment by stsewd** [comment]
Shouldn't this have a data migration for old redirects?

File: readthedocs/redirects/querysets.py [hunk 1]
```diff
@@ -25,7 +30,56 @@ def api(self, user=None, detail=True):
...
+            path__startswith=F('from_url'),
+        )
+        page = Q(
+            redirect_type='page',
+            path__iexact=F('from_url'),
+        )
+        exact = (
+            Q(
+                redirect_type='exact',
+                from_url__endswith='$rest',
```
[the remark this answers is not in the document]
Comment (humitos) [reply readthedocs/redirects/querysets.py]: The downside of this, is that it depends on how the `from_url_without_rest` ends up after the migration and the handling. We strictly depend on setting it correctly.

File: readthedocs/redirects/querysets.py [hunk 2]
```diff
@@ -25,7 +30,56 @@ def api(self, user=None, detail=True):
...
+        )
+        sphinx_htmldir = Q(
+            redirect_type='sphinx_htmldir',
+            path__endswith='.html',
+        )
+
+        queryset = queryset.filter(prefix | page | exact | sphinx_html | sphinx_htmldir)
+
+        # There should be one and only one redirect returned by this query. I
+        # can't think in a case where there can be more at this point. I'm
```
Comment (stsewd) [inline readthedocs/redirects/querysets.py]: A user could create several redirects with the same value, or with `$rest` in different positions, like `/en/latest/foo/$rest` and `/en/latest/foo/bar/$rest`

Comment (stsewd) [reply readthedocs/redirects/querysets.py]: Probably we should sort the result by creation date if we don't do that already

Comment (humitos) [reply readthedocs/redirects/querysets.py]: > like `/en/latest/foo/$rest` and `/en/latest/foo/bar/$rest`

Good point. Although, the first one contains the second one in this case, so it does not make too much sense at this point.

We could implement _new feature_ that the ones that matches _more_ gets priority, though. But that's a different story.

**Pull request merged**